src/algebra/module/torsion.lean

```diff
@@ -0,0 +1,210 @@
+import group_theory.torsion
+
+/-!
+# Torsion submodules
+
+## Main definitions
+
+* `submodule.torsion_by R M a` : the `a`-torsion submodule, containing all elements `x` of `M` such
+  that `a • x = 0`.
+* `submodule.torsion' R M S` : the `S`-torsion submodule, containing all elements `x` of `M` such
+  that `a • x = 0` for some `a` in `S`.
+* `submodule.torsion R M` : the torsion submoule, containing all elements `x` of `M` such that
+  `a • x = 0` for some non-zero-divisor `a` in `R`.
+* `module.is_torsion R M` : the property that defines a torsion module.
+
+## Main statements
+
+* `torsion' R M S` and `torsion R M` are submodules.
+* `torsion_by R M a` can be viewed as a `(R ⧸ R∙a)`-module.
+* `submodule.torsion_by_torsion_by_eq_top` : the `a`-torsion submodule of the `a`-torsion submodule
+  (viewed as a module) is the full `a`-torsion module. Similar lemmas for `torsion'` and `torsion`.
+* `submodule.no_zero_smul_divisors_iff_torsion_bot` : a module over a domain has
+  `no_zero_smul_divisors` (that is, there is no non-zero `a`, `x` such that `a • x = 0`)
+  iff its torsion submodule is trivial.
+* `submodule.quotient_torsion.torsion_eq_bot` : quotienting by the torsion submodule makes the
+  torsion submodule of the new module trivial. If `R` is a domain, we can derive an instance
+  `submodule.quotient_torsion.no_zero_smul_divisors : no_zero_smul_divisors R (M ⧸ torsion R M)`.
+
+## Notation
+
+* The notions are defined for a `comm_semiring R` and a `module R M`. Some additional hypotheses on
+  `R` and `M` are required by some lemmas.
+* The letters `a`, `b`, ... are used for scalars (in `R`), while `x`, `y`, ... are used for vectors
+  (in `M`).
+
+## Tags
+
+Torsion, submodule, module, quotient
+-/
+
+open_locale non_zero_divisors
+section defs
+variables (R M : Type*) [comm_semiring R] [add_comm_monoid M] [module R M]
+namespace submodule
+/-- The `a`-torsion submodule for `a` in `R`, containing all elements `x` of `M` such that
+  `a • x = 0`. -/
+@[simps] def torsion_by (a : R) : submodule R M := (distrib_mul_action.to_linear_map _ _ a).ker
+
+/-- The `S`-torsion submodule, containing all elements `x` of `M` such that `a • x = 0` for some
+`a` in `S`. -/
+@[simps] def torsion' (S : Type*)
+  [comm_monoid S] [distrib_mul_action S M] [smul_comm_class S R M] :
+  submodule R M :=
+{ carrier := { x | ∃ a : S, a • x = 0 },
+  zero_mem' := ⟨1, smul_zero _⟩,
+  add_mem' := λ x y ⟨a, hx⟩ ⟨b, hy⟩,
+    ⟨b * a,
+      by rw [smul_add, mul_smul, mul_comm, mul_smul, hx, hy, smul_zero, smul_zero, add_zero]⟩,
+  smul_mem' := λ a x ⟨b, h⟩, ⟨b, by rw [smul_comm, h, smul_zero]⟩ }
+
+/-- The torsion submodule, containing all elements `x` of `M` such that  `a • x = 0` for some
+  non-zero-divisor `a` in `R`. -/
+@[reducible] def torsion := torsion' R M R⁰
+end submodule
+
+/-- A torsion module is a module whose torsion submodule is the full space. -/
+@[reducible] def module.is_torsion := submodule.torsion R M = ⊤
```

Comment:
I think the definition (outside of this file) will be easier to use if it is
```suggestion
@[reducible] def module.is_torsion : Prop :=
∀ ⦃x⦄, ∃ a : S, a • x = 0
```

And then you can prove `torsion R M = ⊤ ↔ is_torsion R M` as a lemma.

Reply:
Done and duplicated for `torsion_by` and `torsion'`